forester/src/slot_tracker.rs

```diff
@@ -0,0 +1,108 @@
+        Ok(Self {
+            last_known_slot: slot,
+            last_update_time: Instant::now(),
+            update_interval: Duration::from_secs(30),
+            slot_duration: Duration::from_millis(400),
+        })
+    }
+
+    pub async fn update<R: RpcConnection>(&mut self, rpc: &mut R) -> crate::Result<()> {
+        let new_slot = rpc.get_slot().await?;
+        let now = Instant::now();
+        let elapsed = now.duration_since(self.last_update_time);
+        let expected_slots = elapsed.as_millis() as u64 / self.slot_duration.as_millis() as u64;
+        let actual_slots = new_slot.saturating_sub(self.last_known_slot);
+
+        if actual_slots > expected_slots + 1 || expected_slots > actual_slots + 1 {
+            warn!(
+                "Slot drift detected. Expected {} slots, actual {} slots. Resynchronizing.",
+                expected_slots, actual_slots
+            );
+        }
+
+        self.last_known_slot = new_slot;
+        self.last_update_time = now;
+        Ok(())
+    }
+
+    pub fn estimated_current_slot(&self) -> u64 {
+        let elapsed = self.last_update_time.elapsed();
+        let estimated_slots = elapsed.as_millis() as u64 / self.slot_duration.as_millis() as u64;
+        self.last_known_slot + estimated_slots
+    }
+
+    pub async fn run<R: RpcConnection + Send + 'static>(arc_self: Arc<Mutex<Self>>, rpc: &mut R) {
+        let update_interval = {
+            let self_guard = arc_self.lock().await;
+            self_guard.update_interval
+        };
+        let mut interval = interval(update_interval);
+        loop {
+            interval.tick().await;
+            let mut tracker = arc_self.lock().await;
+            if let Err(e) = tracker.update(rpc).await {
+                warn!("Failed to update slot: {:?}", e);
+            }
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub async fn wait_for_slot(target_slot: u64, current_slot: u64, slot_time_ms: u64) {
+    if current_slot < target_slot {
+        let slots_to_wait = target_slot - current_slot;
+        let sleep_ms = slot_time_ms * slots_to_wait;
+        debug!("Sleeping for {} ms", sleep_ms);
+        sleep(Duration::from_millis(sleep_ms)).await;
+    }
+}
+
+pub async fn wait_until_slot_reached<R: RpcConnection>(
+    rpc: &mut R,
+    slot_tracker: &Arc<Mutex<SlotTracker>>,
+    target_slot: u64,
+) -> crate::Result<()> {
+    loop {
+        let current_slot = {
+            let tracker = slot_tracker.lock().await;
+            tracker.estimated_current_slot()
+        };
+
+        if current_slot >= target_slot {
+            let actual_slot = rpc.get_slot().await?;
+            if actual_slot >= target_slot {
+                break;
+            } else {
+                warn!("Slot estimation was ahead. Continuing to wait for actual slot to catch up.");
+            }
+        }
+
+        let slots_to_wait = target_slot.saturating_sub(current_slot);
+        let wait_duration = Duration::from_millis(slots_to_wait * 400);
```

Comment:
can we use `tracker.slot_time` or another unified constant instead of `400` here?